driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -453,13 +470,17 @@ private static void attach(final RetryState retryState, final BulkWriteTracker t
             }
         }
 
-        private BulkWriteTracker(final boolean retry, @Nullable final BulkWriteBatch batch) {
+        private BulkWriteTracker(final boolean retry, @Nullable final BulkWriteBatch batch, final TimeoutContext timeoutContext) {
             attempt = 0;
             attempts = retry ? RetryState.RETRIES + 1 : 1;
             this.batch = batch;
+            this.timeoutContext = timeoutContext;
         }
 
         boolean lastAttempt() {
+            if (timeoutContext.hasTimeoutMS()){
+                return timeoutContext.hasExpired();
+            }
```

Comment:
This change makes lastAttempt non-deterministic within a particular attempt. The following code calls lastAttempt:
```
if (currentBulkWriteTracker.lastAttempt()) {
    addRetryableWriteErrorLabel(writeConcernBasedError, maxWireVersion);
    addErrorLabelsToWriteConcern(result.getDocument("writeConcernError"), writeConcernBasedError.getErrorLabels());
} 
```
The intent is to add error labels on the last attempt. However, now, the above lastAttempt check can return "false", because the timeout has not yet timed out. So labels are not attached. At the end of the iteration, other code will check lastAttempt again, but now the timer will have run out, so no further iterations are performed, and labels are never attached. This appears to be a bug.

Reply:
Thank you for highlighting this issue. Indeed, there was non-deterministic behavior in how `lastAttempt` was handled, which we addressed in a recent update [here](https://github.com/mongodb/mongo-java-driver/pull/1418). With that change, we eliminated the race condition by ensuring `lastAttempt` consistently returns false when `timeoutMS` is present. However, following this update, error labels are never attached when CSOT is enabled. We need to investigate whether we can attach them on each iteration. I will create a ticket to examine this issue before our GA release.